File: lib/schema-format.ts
import type { SchemaTable, SchemaColumn } from './types';
import { quotePgTable, quotePgColumn } from './pg-identifiers';

/**
 * Format the introspected schema into a descriptive string for the AI prompt.
 *
 * For PostgreSQL, every table name and every column name is shown already wrapped
 * in double-quotes so the AI can copy-paste them verbatim into the generated SQL
 * without any case transformation.
 */
export function formatSchemaForPrompt(tables: SchemaTable[], dbType: string): string {
  if (tables.length === 0) {
    return 'No tables found in schema.';
  }

  const isPg = dbType === 'postgresql';

  const lines = tables.map((table) => {
    // Show the identifier exactly as the AI must write it in SQL.
    const tableRef = isPg ? quotePgTable(table.name) : table.name;

    const cols = table.columns
      .map((c: SchemaColumn) => {
        const colRef = isPg ? quotePgColumn(c.name) : c.name;
        const flags: string[] = [];
        if (c.primaryKey) flags.push('PK');
        if (c.autoIncrement) flags.push('auto');
        if (!c.nullable) flags.push('NOT NULL');
        if (c.unique && !c.primaryKey) flags.push('UNIQUE');
        if (c.foreignKey) flags.push(`FK→${isPg ? quotePgTable(c.foreignKey.table) : c.foreignKey.table}.${isPg ? quotePgColumn(c.foreignKey.column) : c.foreignKey.column}`);
        const flagStr = flags.length ? ` [${flags.join(', ')}]` : '';
        return `  - ${colRef} (${c.type})${flagStr}`;
      })
      .join('\n');

    return `Table ${tableRef} (${table.rowCount} rows):\n${cols}`;
  });

  const pgNote = isPg
    ? `\n\nCRITICAL POSTGRESQL RULE: Every table name and column name shown above is already wrapped in double-quotes. You MUST copy them VERBATIM into your SQL — including the double-quotes. Never lowercase, never remove quotes. Example: SELECT "userId", "createdAt" FROM "Users" WHERE "isActive" = true`
    : '';

  return `Database schema (exact identifier casing — do NOT alter):\n\n${lines.join('\n\n')}${pgNote}`;
}

/**
 * Build a simple SELECT * preview for a single table, using properly quoted identifiers.
 */
export function buildSelectPreview(table: SchemaTable, dbType: string, limit = 10): string {
  const tableRef = dbType === 'postgresql' ? quotePgTable(table.name) : table.name;
  return `SELECT * FROM ${tableRef} LIMIT ${limit}`;
}

File: backend/schema_format.py
from typing import List, Dict, Any
from schema import quote_pg_identifier

def format_schema_for_prompt(tables: List[Dict[str, Any]], db_type: str) -> str:
    if not tables:
        return 'No tables found in schema.'

    is_pg = db_type == 'postgresql'
    lines = []

    for table in tables:
        tname = table.get("name", "")
        tref = quote_pg_identifier(tname) if is_pg else tname

        cols = []
        for c in table.get("columns", []):
            cname = c.get("name", "")
            cref = quote_pg_identifier(cname) if is_pg else cname
            ctype = c.get("type", "")
            
            flags = []
            if c.get("primaryKey"):
                flags.append('PK')
            if c.get("autoIncrement"):
                flags.append('auto')
            if not c.get("nullable"):
                flags.append('NOT NULL')
            if c.get("unique") and not c.get("primaryKey"):
                flags.append('UNIQUE')
            
            fk = c.get("foreignKey")
            if fk:
                fk_table = quote_pg_identifier(fk.get("table")) if is_pg else fk.get("table")
                fk_col = quote_pg_identifier(fk.get("column")) if is_pg else fk.get("column")
                flags.append(f'FK→{fk_table}.{fk_col}')
                
            flag_str = f" [{', '.join(flags)}]" if flags else ""
            cols.append(f"  - {cref} ({ctype}){flag_str}")
            
        cols_str = "\n".join(cols)
        row_count = table.get("rowCount", 0)
        lines.append(f"Table {tref} ({row_count} rows):\n{cols_str}")

    pg_note = (
        "\n\nCRITICAL POSTGRESQL RULE: Every table name and column name shown above is already wrapped in double-quotes. "
        "You MUST copy them VERBATIM into your SQL — including the double-quotes. Never lowercase, never remove quotes. "
        "Example: SELECT \"userId\", \"createdAt\" FROM \"Users\" WHERE \"isActive\" = true"
        if is_pg else ""
    )

    return f"Database schema (exact identifier casing — do NOT alter):\n\n{'\n\n'.join(lines)}{pg_note}"

def build_select_preview(table: Dict[str, Any], db_type: str, limit: int = 10) -> str:
    tname = table.get("name", "")
    tref = quote_pg_identifier(tname) if db_type == 'postgresql' else tname
    return f"SELECT * FROM {tref} LIMIT {limit}"
